src/DataStructures/ComplexDataVector.hpp

```diff
@@ -0,0 +1,179 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines class Data.
+
+#pragma once
+
+#include <complex>
+#include <cstddef>
+#include <functional>  // for std::reference_wrapper
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/VectorImpl.hpp"
+#include "Utilities/ConstantExpressions.hpp"
+#include "Utilities/ForceInline.hpp"
+#include "Utilities/Requires.hpp"
+#include "Utilities/StdArrayHelpers.hpp"
+
+// IWYU pragma: no_forward_declare ConstantExpressions_detail::pow
+
+/*!
+ * \ingroup DataStructuresGroup
+ * \brief Stores a collection of complex function values.
+ *
+ * \details Use ComplexDataVector to represent function values on the
+ * computational domain. Note that interpreting the data also requires knowledge
+ * of the points that these function values correspond to.
+ *
+ * A ComplexDataVector holds an array of contiguous data. The ComplexDataVector
+ * can be owning, meaning the array is deleted when the ComplexDataVector goes
+ * out of scope, or non-owning, meaning it just has a pointer to an array.
+ *
+ * Refer to the \ref DataStructuresGroup documentation for a list of other
+ * available types.
+ *
+ * ComplexDataVectors support a variety of mathematical operations that are
+ * applicable to nodal coefficients. In addition to common arithmetic operations
+ * such as elementwise addition, subtraction, multiplication and division,
+ * elementwise operations on blaze vectors of doubles are supported. See
+ * [blaze-wiki/Vector_Operations]
+ * (https://bitbucket.org/blaze-lib/blaze/wiki/Vector%20Operations).
+ *
+ * in addition, support is provided for:
+ * - abs  (returns DataVector, as the result is real)
+ * - conj
+ * - imag  (returns DataVector, as the result is real)
+ * - real  (returns DataVector, as the result is real)
+ */
+class ComplexDataVector
+    : public VectorImpl<std::complex<double>, ComplexDataVector> {
+ public:
+  using VectorImpl<std::complex<double>, ComplexDataVector>::operator=;
+  using VectorImpl<std::complex<double>, ComplexDataVector>::VectorImpl;
+
+  MAKE_MATH_ASSIGN_EXPRESSION_ARITHMETIC(ComplexDataVector)
+};
+
+namespace blaze {
+VECTOR_BLAZE_TRAIT_SPECIALIZE_ARITHMETIC_TRAITS(ComplexDataVector);
+VECTOR_BLAZE_TRAIT_SPECIALIZE_ALL_MAP_TRAITS(ComplexDataVector);
+template <>
+struct UnaryMapTrait<ComplexDataVector, blaze::Real> {
+  using Type = DataVector;
+};
+template <>
+struct UnaryMapTrait<ComplexDataVector, blaze::Imag> {
+  using Type = DataVector;
+};
+template <>
+struct UnaryMapTrait<DataVector, blaze::Real> {
+  using Type = DataVector;
+};
+template <>
+struct UnaryMapTrait<DataVector, blaze::Imag> {
```

Comment:
what does this do?

Reply:
It should always return 0, but this should be present for compatibility with the `std::imag`, which returns 0 for floating-point fundamentals.

Reply:
okay